echopype/echodata/echodata.py

```diff
@@ -1,57 +1,455 @@
+            out_dir = current_dir.joinpath(Path("temp_echopype_output"))
+            if not out_dir.exists():
+                out_dir.mkdir(parents=True)
+
+            warnings.warn(f"Resulting converted file(s) will be available at {str(out_dir)}")
+            out_path = [
+                str(out_dir.joinpath(Path(os.path.splitext(Path(f).name)[0] + file_format)))
+                for f in source_file
+            ]
+
+        else:
+            fsmap = fsspec.get_mapper(save_path, **output_storage_options)
+            output_fs = fsmap.fs
+
+            # Use the full path such as s3://... if it's not local, otherwise use root
+            if isinstance(output_fs, LocalFileSystem):
+                root = fsmap.root
+            else:
+                root = save_path
+
+            fname, ext = os.path.splitext(root)
+            if ext == "":  # directory
+                out_dir = fname
+                out_path = [
+                    os.path.join(root, os.path.splitext(os.path.basename(f))[0] + file_format)
+                    for f in source_file
+                ]
+            else:  # file
+                out_dir = os.path.dirname(root)
+                if len(source_file) > 1:  # get dirname and assemble path
+                    out_path = [
+                        os.path.join(
+                            out_dir, os.path.splitext(os.path.basename(f))[0] + file_format
+                        )
+                        for f in source_file
+                    ]
+                else:
+                    # force file_format to conform
+                    out_path = [
+                        os.path.join(
+                            out_dir, os.path.splitext(os.path.basename(fname))[0] + file_format
+                        )
+                    ]
+
+        # Create folder if save_path does not exist already
+        fsmap = fsspec.get_mapper(str(out_dir), **output_storage_options)
+        fs = fsmap.fs
+        if file_format == ".nc" and not isinstance(fs, LocalFileSystem):
+            raise ValueError("Only local filesystem allowed for NetCDF output.")
+        else:
+            try:
+                # Check permission, raise exception if no permission
+                io.check_file_permissions(fsmap)
+                if isinstance(fs, LocalFileSystem):
+                    # Only make directory if local file system
+                    # otherwise it will just create the object path
+                    fs.mkdir(fsmap.root)
+            except FileNotFoundError:
+                raise ValueError("Specified save_path is not valid.")
+
+        return out_path  # output_path is always a list
+
+    @staticmethod
+    def _normalize_path(out_f, convert_type, output_storage_options):
+        if convert_type == "zarr":
+            return fsspec.get_mapper(out_f, **output_storage_options)
+        elif convert_type == "netcdf4":
+            return out_f
+
+    def _to_file(
```

Comment:
Hmm... looks like you integrated a lot of the stuff from Convert Object ... Not sure if this is the best way to go because I can see EchoData ending up being convoluted ... I think it might be better if you keep all of these functions in the Convert Object and use that within EchoData.

Reply:
Yeah, those transplanted functions are my biggest hesitation, as in whether to keep them in Convert or move them to EchoData. (I was thinking about getting rid of the Convert object!) But looking at this again maybe it'll be better that way.

Reply:
I see. I think it's okay to keep the Convert object. Since we know it already works and this would split up the functionalities, and we can just test individually.

Reply:
Oh now I remember why I did that: the main thing is that now the set groups functions return xr.Dataset and are linked directly to attributes of EchoData in open_raw. That's why I felt the Convert class can actually be removed and substituted by the same functionalities in EchoData. This does make EchoData more convoluted though I agree. Hmm... thoughts?